libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java

```diff
@@ -209,7 +209,7 @@ public void closeFromSelector() throws IOException {
 
     protected void handleReadBytes() throws IOException {
         int bytesConsumed = Integer.MAX_VALUE;
-        while (bytesConsumed > 0 && channelBuffer.getIndex() > 0) {
+        while (selectorShouldClose() == false && bytesConsumed > 0 && channelBuffer.getIndex() > 0) {
```

Comment:
This feels odd. We are essentially handling bytes that have been read from the channel at the application level here. I do not understand why we would want to stop handling bytes that we have read based on asynchronous signals to close the channel from a different thread. I think this should be removed until it is proven that we need this.

Reply:
Yea, your best guess below is correct, see e.g. this stacktrace:

```
org.elasticsearch.nio.NioSelector.writeToChannel(NioSelector.java:327)
org.elasticsearch.nio.SocketChannelContext.sendMessage(SocketChannelContext.java:143)
org.elasticsearch.transport.nio.MockNioTransport$MockSocketChannel.sendMessage(MockNioTransport.java:314)
org.elasticsearch.transport.OutboundHandler.internalSendMessage(OutboundHandler.java:80)
org.elasticsearch.transport.OutboundHandler.sendMessage(OutboundHandler.java:70)
org.elasticsearch.transport.TcpTransport.sendResponse(TcpTransport.java:738)
org.elasticsearch.transport.TcpTransport.lambda$new$1(TcpTransport.java:162)
org.elasticsearch.transport.TransportHandshaker.handleHandshake(TransportHandshaker.java:100)
org.elasticsearch.transport.TcpTransport.handleRequest(TcpTransport.java:1019)
org.elasticsearch.transport.TcpTransport.messageReceived(TcpTransport.java:922)
org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:753)
org.elasticsearch.transport.TcpTransport.consumeNetworkReads(TcpTransport.java:780)
org.elasticsearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:241)
org.elasticsearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:213)
org.elasticsearch.nio.BytesChannelContext.read(BytesChannelContext.java:47)
org.elasticsearch.nio.EventHandler.handleRead(EventHandler.java:119)
org.elasticsearch.nio.NioSelector.handleRead(NioSelector.java:399)
org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:246)
org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:172)
org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:129)
java.base/java.lang.Thread.run(Thread.java:834)
```

where the post handling hits us.

Reply:
This would also be an example case for the question:

> Can you give some more context about this exception and when it happens?

Where a delete in the post handler of the write then leads to the `channelBuffer` being closed and the pages dequeue being empty.
It's not really a concurrency issue, it's simply the fact that we don't break out on closed channels/contexts.

Reply:
Let's change it to `isOpen()`. I think that makes the most sense for use internal by the context (opposed to ready to close).